src/vehicle_autodrive.cpp

```diff
@@ -1,25 +1,1202 @@
 #include "character.h"
 #include "coordinates.h"
+#include "cuboid_rectangle.h"
+#include "debug.h"
+#include "enums.h"
 #include "game.h"
 #include "map.h"
+#include "map_iterator.h"
+#include "map_memory.h"
+#include "mapdata.h"
+#include "messages.h"
+#include "optional.h"
 #include "point.h"
+#include "tileray.h"
+#include "translations.h"
+#include "type_id.h"
+#include "veh_type.h"
+#include "vpart_position.h"
+
+/*
+ * This file contains code that allows a vehicle to be driven by an in-game character (most
+ * likely the player character) to some remote destination, following a path of adjacent overmap
+ * tile (OMT) coordinates, while avoiding any obstacles along the way.
+ *
+ * The main entry point is do_autodrive(), which is intended to be called from the do_turn() of
+ * a long-running activity. This will cause the driver character to perform one or more driving
+ * actions (as long as they have enough moves), which can be either steering or changing the
+ * cruise control speed. The driver may also do nothing if the vehicle is going in the right
+ * direction.
+ *
+ * Most of the logic is inside the private implementation class autodrive_controller, which is
+ * instantiated at the beginning of an autodrive activity and cleaned up at the end. This class
+ * caches information about the vehicle, driver and the local terrain in order to minimize the
+ * amount of computation required on each invocation of do_autodrive().
+ *
+ * Upon reaching a new overmap tile, as well as once initially, the autodrive algorithm will
+ * try to find a path that takes the vehicle from its current position through the current
+ * and next OMT in order to reach the 2nd next OMT in the overall path to destination (which
+ * is obtained from driver.omt_path). Thus we always plan ahead 2 OMTs worth of navigation,
+ * but we only follow the first half of the route, then recompute a new route for the next
+ * 2 OMTs. This ensures that we don't drive up to the edge of an OMT only to get stuck due
+ * to a large obstacle that can't be driven around from that position (a dead end) and instead
+ * we always choose to cross OMT boundaries in a position from which we are guaranteed to be
+ * able to continue onward.
+ *
+ * The pathfinding algorithm being employed is A*. The graph being searched is made up of nodes
+ * denoted by x and y coordinates as well as a direction of travel. Since steering changes the
+ * vehicle's direction in increments of 15 degrees there are 360/15 possible orientations. By
+ * limiting navigation to 2 OMTs at a time we have a maximum search space of 2*24*24*24 nodes.
+ * In practice, A* finds a good path after visiting fewer than 50 nodes due to the heuristic
+ * function that estimates which nodes are most likely to lead to the destination (which is
+ * based on proximity to the "exit zone" leading to the next OMT and whether the vehicle is
+ * pointing towards the middle of that zone). Although there can be many paths that we could
```

Comment:
One could use the prior iteration (second half of the previous route) as a part of this; I don't know if it would be worth the effort of coding it, though. Any data on how well this heuristic is performing?

A second question: If cost_so_far+heuristic is the same for two possible nodes, what determines tiebreaking?

Reply:
It's either the first or the last such node in the order of insertion into the priority queue (not sure which).